ringpop.thrift-gen

```diff
@@ -0,0 +1,171 @@
+    if dest == a.ringpop.WhoAmI() {
+        // use local client
+        return a.impl
+    }
+
+    // create remote client
+    client := thrift.NewClient(
+        a.ch,
+        a.ch.ServiceName(),
+        &thrift.ClientOptions{
+            HostPort: dest,
+        },
+    )
+    return New{{ $TChanInterface }}Client(client)
+}
+
+{{ range .Methods }}
+{{ $Method := .}}
+{{ $PublicName := $Method.Name | goPublicName }}
+{{ $cfg := print "config." $PublicName }}
+{{ $ConfigType := print ($Service.Name | goPublicName) $PublicName "Configuration"}}
+
+// ------ 'Generated' code for {{ $Method.Name | goPublicName }} ------ {{ $cfg }}
+type {{ $ConfigType }} struct {
+    N       ringpop.ReplicationFactor
+    Fanout  ringpop.ReplicationFactor
+    Reducer func({{ if $Method.ReturnType }}rs []{{ $Method.ReturnType | goType }}, {{ end }}errs []error) ({{ template "returnType" $Method}})
+    Key     func({{ template "arguments" $Method}}) (string, error)
+}
+
+func (a *{{ $Adapter }}) {{ $PublicName }} ({{ template "arguments" $Method }}) ({{ template "returnType" $Method }}) {
+    // check if the function is configured to be ring aware
+    if a.{{ $cfg }} == nil {
+        return a.impl.{{ $PublicName }}({{ template "call" $Method }})
+    }
+
+    if hasForwardedHeader(ctx) {
+        return a.impl.{{ $PublicName }}({{ template "call" $Method }})
+    }
+
+    // find the key to shard on
+    ringpopKey, err := a.{{ $cfg }}.Key({{ template "call" $Method }})
+    if err != nil {
+        return {{ if $Method.ReturnType }}r, {{ end }}fmt.Errorf("could not get key: %s", err)
+    }
+
+    clients := a.router.GetClients(ringpopKey, a.{{ $cfg }}.N, a.{{ $cfg }}.Fanout)
+    clientLength := len(clients)
+    if clientLength == 0 {
+        return {{ if $Method.ReturnType }}r, {{ end }}errors.New("no clients")
+    }
+
+    // slices to collect return values
+    {{ if $Method.ReturnType }}
+        rs := make([]{{ $Method.ReturnType | goType }}, clientLength)
+    {{ end }}
+    errs := make([]error, clientLength)
+
+    // make headers indicate a forwarded request
+    ctx = setForwardedHeader(ctx)
+
+    var wg sync.WaitGroup
+    for i, untypedClient := range clients {
+        client := untypedClient.({{ $TChanInterface }})
+        wg.Add(1)
+        go func(i int, client {{ $TChanInterface }}) {
+            defer wg.Done()
+
+            {{ if $Method.ReturnType }}
+                rs[i], errs[i] =
+            {{ else }}
+                errs[i] =
+            {{ end }}
+                client.{{ $PublicName }}({{ template "call" $Method }})
+        }(i, client)
+    }
+    wg.Wait()
+
+    // allow nil Reducer with only 1 client
+    if clientLength == 1 && a.{{ $cfg }}.Reducer == nil {
+        return {{ if $Method.ReturnType }}rs[0], {{ end }}errs[0]
+    }
+    return a.{{ $cfg }}.Reducer({{ if $Method.ReturnType }}rs, {{end}}errs)
+}
+{{ end}}
+// ------ End of {{ $Service.Name }} ------
+{{ end}}
+
+// Forwarding helpers
```

Comment:
Ahh I see. It's probably fine to assume it will be `thrift.Context`

I was considering letting the user choose between `tchannel.ContextWithHeaders` and `thrift.Context`. However, it ended up just being a constant, so you can pretty safely assume it'll be `thrift.Context` I think.